api/base/utils.py

```diff
@@ -70,15 +71,31 @@ def absolute_reverse(view_name, query_kwargs=None, args=None, kwargs=None):
     return url
 
 
-def get_object_or_error(model_cls, query_or_pk, display_name=None, **kwargs):
+def get_object_or_error(model_cls, query_or_pk, display_name=None, prefetch_fields=list()):
+    obj = query = None
     if isinstance(query_or_pk, basestring):
-        obj = model_cls.load(query_or_pk)
-        if obj is None:
-            raise NotFound
+        if issubclass(model_cls, GuidMixin):
+            query = {'guids___id': query_or_pk}
+        else:
+            if hasattr(model_cls, 'primary_identifier_name'):
+                query = {model_cls.primary_identifier_name: query_or_pk}
+            else:
+                obj = model_cls.load(query_or_pk)
     else:
+        if hasattr(model_cls, 'primary_identifier_name'):
+            query = to_django_query(query_or_pk, model_cls=model_cls)
+        else:
+            obj = model_cls.find_one(query_or_pk)
+
+    if not obj:
```

Comment:
If we could find a way around the metaclass magic in FileNode this wouldn't be necessary. Just time constraints what they are I'm copping out.